### ISSUE TITLE
Latest: SAS: Simple Attention Sparsification via End-to-End Optimization of Context Ranking

### ISSUE THREAD

**Issue by jjakimoto** [jjakimoto/research-issues#1534]
**Topic:** Quantitative trading; machine learning for quant & trading; agentic coding techniques; LLMs & multimodal LLMs; and autonomous research (autoresearch / AI-scientist) systems — methods, papers, tools  
**Date:** 2026-09-15

**Source:** huggingface-daily-papers · 2026-09-11
**Link:** https://arxiv.org/abs/2609.13141

SAS shows you can train a sparse-attention block selector directly with the language modeling loss instead of distilling dense attention maps — one gated Triton kernel swap — and it recovers near-full-attention quality at a 4096-token KV budget while decoding up to 5.6x faster at 512K context on a single GPU.

## Executive summary

SAS (Tencent Hunyuan + HKUST) attacks post-training attention sparsification: pretrained dense Transformers are adapted so each query attends only to a Top-K set of context blocks. Existing trainable selectors (SeerAttention-R, SeerAttention) cannot backprop through hard Top-K, so they are supervised by layer-wise distillation of the dense attention distribution — a surrogate that ranks blocks by where the dense model attends rather than by what actually helps the final prediction. SAS instead injects the selector's softmax-normalized scores as log-space gates inside the attention softmax during training, making the selector differentiable under the plain LM loss, and implements it as a FlashAttention-style fused Triton kernel so no gated attention matrix is ever materialized. On Qwen3-4B/8B/14B it beats SeerAttention-R at every budget, with the biggest gains at tight budgets, and its SGLang decode backend is 5.6x faster than dense at 512K (batch 1) and 13x at 64K (batch 8).

### The problem: gradient blockage and ranking misalignment

Block sparse attention partitions the context into blocks of B tokens, has a lightweight selector score them per query, and attends only to the Top-K. Because the selected index set is piecewise constant in the selector scores, the LM loss gives zero gradient through the selection path. Trainable methods therefore distill the dense teacher's per-layer attention mass into the selector. The authors argue this is the wrong objective on two counts: per-layer targets ignore cross-layer complementarity (a block missed at one layer can be picked up at another), and attention matching constrains only the attention weights, ignoring how the attended V values shape the final prediction.

- Setting: backbone frozen, only the selector trained (controlled comparison vs distillation).
- Built on SeerAttention-R's open-source AttnGate selector; identical architecture, identical data — only the training signal differs.
- Block size 64, Top-K 32 (i.e. a 2048-token budget in the ablations).

### The mechanism: log-space gates inside the softmax

The always-retained current block gets a unit gate (zero additive bias). Historical block scores are turned into positive gates via a softmax over blocks and added in log form to the attention logits before the softmax, so the gates modulate how attention mass is allocated rather than merely rescaling the output. At inference the learned ranking is simply discretized back to Top-K, so the inference path is unchanged from SeerAttention-R — only the gate checkpoint differs.

- Inner gate (inside softmax) vs outer gate (rescaling the output) is the single biggest design lever.
- Softmax normalization calibrates historical context against the unit-gated current block and makes gates invariant to a global logit shift.
- No teacher model, no auxiliary distillation loss, no extra hyperparameter balancing.

### Four design choices, ablated

A controlled ablation on GPQA-Diamond (Qwen3-4B, 2048 budget, avg@16 over one epoch of OpenR1-Math-220k) isolates gate position, gate activation, ranking preservation, and training scope. Inner+softmax+soft+sparse is the winning recipe. Sigmoid gates saturate toward 1 and raw logit injection collapses toward 0 with shrinking variance — both degenerate to effectively ungated attention, which is why the competitive softmax matters. Hard Top-K gating with a straight-through estimator is unstable: its forward normalizer sums only over the selected set, so a dropped block that outscores the selected set gets an unbounded weight and gradient norms orders of magnitude larger than soft gating.

- Inner gating: 54.4 vs outer 41.6 after one epoch.
- Softmax activation: 54.4 vs sigmoid 17.0 vs raw-logit injection 18.8 — a collapse, not a small regression.
- Soft (continuous) gates: 54.4 vs hard STE 46.0.
- Training scope barely matters for final quality (sparse 54.8 vs full 54.4), so they use the cheaper sparse scope.

### Kernel and serving story

A naive implementation would materialize the gated attention matrix, which is prohibitive at 32K training length. Their Triton kernel fuses the per-block log gate into the FlashAttention tile scan: during the streaming key-value pass it adds the log gate to the logits of selected historical blocks, masks non-selected ones, leaves the current block unbiased, and does the standard online softmax update; the backward accumulates block-level gate gradients by summing attention-logit gradients within each selected block. Top-K is realized without an explicit sort via a per-query threshold on the gate. Inference is a native SGLang attention backend on paged KV + FlashInfer sparse decode, with dense prefill and sparse decode, GQA per-group block selection, CUDA-graph capture, and batched decode.

- Decode complexity drops from O(L) per step to O(K*B), independent of total context length.
- Training: AdamW lr 1e-3, cosine, global batch 32, 32,768 max sequence length, one epoch, VeOmni + FSDP.
- One selector trained on math data only, reused unchanged for reasoning, LongBench, and agentic evals.

### Why it wins: cross-layer complementarity, not more attention mass

Counterintuitively, SAS covers LESS per-layer attention mass than the distilled selector in almost all layers, under both a raw-weight metric and a value-magnitude-aware metric. But when you take the union of selected blocks across all layers and measure overlap recall against a full-attention oracle (majority-voted per GQA group on 8K sequences with Qwen3-4B), SAS is consistently higher. Distillation optimizes a local, per-layer objective; end-to-end training lets layers specialize and cover the oracle jointly. A second-order effect: better selection means shorter reasoning traces and fewer truncations at a 32,768 max generation length, with the gap largest on AIME.

- Less mass per layer, higher cross-layer union recall — a useful mental model for anyone tuning KV selection.
- Token efficiency, not just accuracy, improves at a 4096 budget on Qwen3-4B.

### Decode efficiency and the new bottleneck

Steady-state decode on Qwen3-4B in SGLang, single GPU, CUDA graphs, 512 generated tokens timed after prefill. Dense latency grows linearly with context; SAS stays nearly flat. Near parity at 8K, then 2.4x / 4.6x / 5.6x faster at 64K / 256K / 512K at batch 1; 13x at 64K with batch 8. Tighter budgets are only modestly faster, so the win comes from avoiding the full KV read, not from the exact K. Breaking one sparse decode step into selector scoring, Top-K selection, and sparse attention compute shows the attention compute is context-invariant while selector scoring scans all block summaries and Top-K must rank all candidates: Top-K goes from 21% of the step at 8K (where attention compute is 58%) to 90% at 512K. The selection stage, not the attention, is the thing to optimize next.

- Prefill is unchanged — SAS sparsifies decode only.
- Practical note for local long-context serving: the payoff starts above ~8K and compounds with batch size.

**Reasoning benchmarks, Qwen3 (subset of Table 2; accuracy %)**

| Budget | Method | MATH500 4B | MATH500 14B | GPQA-D 4B | GPQA-D 14B | AIME24 4B | AIME25 4B |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Full | Full Attn | 93.93 | 95.22 | 56.19 | 65.25 | 71.25 | 66.41 |
| 1024 | StreamingLLM | 72.72 | 73.47 | 13.86 | 19.07 | - | - |
| 1024 | SeerAttn-R | 84.67 | 86.12 | 39.84 | 45.64 | - | - |
| 1024 | SAS | 90.65 | 92.93 | 50.41 | 61.14 | - | - |
| 2048 | Quest | 40.74 | 81.52 | 12.17 | 41.01 | 0 | 0 |
| 2048 | SeerAttn-R | 91.85 | 93.02 | 49.94 | 61.68 | 55.83 | 45.16 |
| 2048 | SAS | 93.47 | 93.54 | 54.86 | 65.09 | 68.85 | 56.38 |
| 4096 | SeerAttn-R | 94.10 | 95.12 | 55.40 | 63.83 | 69.32 | 58.59 |
| 4096 | SAS | 93.67 | 95.23 | 55.11 | 65.03 | 71.72 | 59.97 |

**LongBench by input length, Qwen3-14B (Table 3)**

| Budget | Method | 0-4K | 4-8K | 8K+ | Avg. |
| --- | --- | --- | --- | --- | --- |
| Full | Full Attn | 59.2 | 55.6 | 55.0 | 56.6 |
| 2048 | SeerAttn-R | 58.9 | 54.3 | 51.5 | 54.9 |
| 2048 | SAS | 59.2 | 54.4 | 53.9 | 55.8 |
| 4096 | SeerAttn-R | 59.0 | 54.8 | 53.5 | 55.7 |
| 4096 | SAS | 58.9 | 54.8 | 54.8 | 56.2 |

**Gating ablation on GPQA-Diamond, Qwen3-4B, 2048 budget, avg@16 after 1 epoch (Table 1)**

| Setting | Configuration | 10 step | 100 step | 1000 step | 1 epoch |
| --- | --- | --- | --- | --- | --- |
| (1) | Baseline (full attention) | - | - | - | 56.1 |
| (2) | inner gate, softmax, soft, full scope | 30.8 | 52.2 | 53.7 | 54.4 |
| (3) | outer gate, softmax, soft, full scope | 28.3 | 38.9 | 43.2 | 41.6 |
| (5) | inner, sigmoid, soft, full scope | 19.6 | 20.4 | 17.6 | 17.0 |
| (6) | inner, raw-logit injection, soft, full scope | 23.4 | 19.9 | 18.9 | 18.8 |
| (8) | inner, softmax, hard Top-K + STE, full scope | 46.5 | 42.4 | 49.6 | 46.0 |
| (10) | inner, softmax, soft, sparse scope (SAS) | 24.8 | 51.0 | 53.6 | 54.8 |

**BFCL Multi-Turn (Table 4)**

| Budget | Method | Qwen3-4B | Qwen3-8B | Qwen3-14B |
| --- | --- | --- | --- | --- |
| Full | Full Attn | 35.75 | 42.75 | 44.50 |
| 2048 | SeerAttn-R | 29.00 | 34.38 | 38.88 |
| 2048 | SAS | 32.50 | 36.75 | 39.50 |
| 4096 | SeerAttn-R | 33.38 | 39.25 | 43.88 |
| 4096 | SAS | 34.38 | 41.25 | 44.00 |

## Key findings

- At a 1024-token budget SAS beats distillation-trained SeerAttention-R by 6.0-7.7 points on MATH500 and 10.6-15.5 points on GPQA-Diamond across Qwen3-4B/8B/14B; the tighter the budget, the larger the gap.
- At a 2048 budget on Qwen3-4B, SAS gains +13.0 on AIME24 over SeerAttention-R (68.85 vs 55.83) with an identical AttnGate selector and identical data — the only difference is the training signal.
- At a 4096 budget SAS matches or exceeds dense attention on several reasoning tasks (AIME24 Qwen3-4B: 71.72 vs 71.25 full) and essentially recovers full attention on LongBench (56.2 vs 56.6 avg on Qwen3-14B).
- Gate design is brittle in specific ways: softmax-normalized inner gates reach 54.4 on GPQA-D while sigmoid gates collapse to 17.0 and raw-logit injection to 18.8; hard Top-K with STE reaches only 46.0 due to unbounded weights on dropped blocks.
- SGLang decode: near parity at 8K but 2.4x / 4.6x / 5.6x faster than dense at 64K / 256K / 512K at batch 1, and 13x at 64K at batch 8.
- Top-K ranking, not attention compute, becomes the decode bottleneck at long context — 21% of the step at 8K rising to 90% at 512K.
- SAS selects LESS per-layer attention mass than distillation yet achieves higher cross-layer union overlap recall against a full-attention oracle, and generates shorter traces with a lower truncation rate at 32,768 max generation length.

## Notable quotes

> At budget 4096, SAS matches or exceeds full attention in several cases (e.g., 71.72 vs. 71.25 on AIME24, Qwen3-4B), recovering full attention reasoning while attending to only a small fraction of the KV blocks.
> On BFCL multi-turn (Table 4), SAS beats SeerAttention-R across all backbones and budgets, e.g., +3.5 on Qwen3-4B at budget 2048, and nearly closes the gap to full attention at budget 4096 (e.g., 44.00 vs. 44.50 on Qwen3-14B).
> As a result, even though SAS retains less attention mass within each individual layer, its per-layer choices are more complementary, and their union attains higher overlap recall against the oracle.
> Across all four reasoning benchmarks, SAS produces shorter generations and a lower truncation rate, with the gap widening on the harder AIME tasks where long reasoning is required.

## Limitations

- Appendix B: on RULER the selector degrades substantially as context grows and leaves a considerable gap from full attention even after training on 64K sequences with 0.5B tokens; the authors attribute this to pooling-based block summaries losing fine-grained, needle-like signals.
- All post-training experiments freeze the backbone and train only the selector on a single math dataset (OpenR1-Math-220k), so long-context and agentic transfer is measured out of distribution; the continued-pretraining result is on one 7B model and only about matches a plain sliding-window baseline on downstream average (43.28 vs 43.24).
- Decode speedups are measured on a single GPU with one model (Qwen3-4B) in SGLang, and prefill is left dense, so end-to-end serving gains for prefill-heavy workloads will be smaller.

## Open questions

- Can a more expressive but still cheap block summary (beyond pooling) close the RULER/needle gap at 128K+?
- Since Top-K ranking dominates the decode step at 512K, what selection kernel or hierarchical/approximate ranking removes that bottleneck?
- Does end-to-end gated training hold up when the backbone is also trained at larger scale, and does it compose with RL post-training rather than just supervised LM loss?

---
_Generated by Auto Research via a headless coding agent. Sources were retrieved with live web search — please verify before citing._

**Comment by x711criptic** [comment]
Might be relevant here — x711 is a pay-per-use tool API for autonomous agents. 41 tools (web search, crypto prices, shared Hive memory, on-chain tx sim, email, agent spawning), free tier 10 calls/day:

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"web_search","query":"your query"}'
```

Works via REST or MCP. 1,000+ agents on it across LangChain, CrewAI, AutoGen, ElizaOS.

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by jjakimoto** [comment]
Superseded by a fresh Auto Research run — closing to reset the slate.
